### PR TITLE
UserStory/#37 make an online appointment

### LINKED ISSUE #37: Make an online appointment
As a Customer, I want to make appointments, so I can get services.

### Acceptance Criteria

- Choose saloon from list of salons.
- Window with services is opened.
- _Checkbox_ is located near the service`s name. 
- When customer clicks on the one or more checkboxes selecting it, he or she  chooses service near it.  
- When customer chooses one or more services on the bottom of them button _"Add date"_  appears.
- When customer clicks on it new window opens with calendar.
 - Clicking on date from calendar new window opens with time slots.
 - Clicking on time slot on page customer chooses  time for appointment.
 - Customer can add additional note into the field _"Extra info"_  under calendar, but it is not necessarily.
- After choosing date and time on the bottom of it button _"Add order"_  is located when click on it order is created.
- Customer is redirecting to the own online cabinet where is his created order over there but not active.
- Specialist  receives a email message that he needs to confirm new order.
- Specialist opens own online cabinet with new orders one of them is ours.
- 2  button are near order: approve and cancel, Specialist clicks on the one of them. If Specialist confirms the order it becomes active, otherwise it becomes canceled and delete from list.
- Email message is sent to customer from site  with  a  decision of specialist.
- Then customer refreshes online cabinet page.
- If specialist approves this order becomes active else it deletes from cabinet.

- [x] #141
- [x] #140
- [x] #142
- [x] #143

### PR DESCRIPTION
**Pull request by Misha86**
- [x] #141
- [x] #140
- [x] #142
- [x] #143

### CODE REVIEW

File: beauty/api/apps.py [hunk 1]
```diff
@@ -4,3 +4,9 @@
 class ApiConfig(AppConfig):
     default_auto_field = 'django.db.models.BigAutoField'
```
Comment (Serhii-Voloshyn) [inline beauty/api/apps.py]: Docstring needed.

File: beauty/api/apps.py [hunk 2]
```diff
@@ -4,3 +4,9 @@
...
     name = 'api'
+
+    def ready(self):
+        # Implicitly connect a signal handlers decorated with @receiver.
+        from beauty import signals
+        # Explicitly connect a signal handler.
+        # signals.request_finished.connect(signals.my_callback)
```
Comment (Serhii-Voloshyn) [inline beauty/api/apps.py]: If you don't need this code -- delete it.

File: beauty/api/models.py [hunk 3]
```diff
@@ -588,7 +590,6 @@ def __str__(self):
         return self.name
 
     class Meta:
-
         """This meta class stores ordering and verbose name"""
```
Comment (Serhii-Voloshyn) [inline beauty/api/models.py]: Blank line after class docstring.

File: beauty/api/serializers/business_serializers.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+from rest_framework import serializers
+from api.models import Business, CustomUser
+
+
+class BusinessListCreateSerializer(serializers.ModelSerializer):
```
Comment (Serhii-Voloshyn) [inline beauty/api/serializers/business_serializers.py:5]: Add docstring and blank line

File: beauty/api/serializers/business_serializers.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    class Meta:
```
Comment (Serhii-Voloshyn) [inline beauty/api/serializers/business_serializers.py:6]: Add docstring and blank line

File: beauty/api/serializers/order_serializers.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
+"""The module includes serializers for Order model."""
+
+from rest_framework import serializers
+from api.models import (Order, CustomUser, Service, Position)
+import logging
+
```
Comment (Serhii-Voloshyn) [inline beauty/api/serializers/order_serializers.py:6]: 2 blank lines after import section

File: beauty/api/tests/test_order_serializers.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+from django.utils import timezone
+from django.test import TestCase
+from rest_framework.exceptions import ErrorDetail, ValidationError
+
+from api.serializers.order_serializers import OrderSerializer
+from .factories import (GroupFactory,
+                        CustomUserFactory,
+                        PositionFactory,
+                        ServiceFactory)
+
```
Comment (Serhii-Voloshyn) [inline beauty/api/tests/test_order_serializers.py:26]: 2 blank lines after import section

File: beauty/api/tests/test_order_views.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+from api.serializers.order_serializers import OrderSerializer
+from api.views import (OrderListCreateView,
+                       OrderApprovingTokenGenerator)
+from .factories import (GroupFactory,
+                        CustomUserFactory,
+                        PositionFactory,
+                        ServiceFactory,
+                        OrderFactory)
+
```
Comment (Serhii-Voloshyn) [inline beauty/api/tests/test_order_views.py:35]: 2 blank lines after import section

File: beauty/api/views.py [hunk 9]
```diff
@@ -1,78 +1,101 @@
...
+from .permissions import (IsAccountOwnerOrReadOnly, IsOrderUserOrReadOnly)
+
+from .serializers.customuser_serializers import (CustomUserDetailSerializer,
+                                                 CustomUserSerializer,
+                                                 ResetPasswordSerializer)
+from api.serializers.order_serializers import (OrderSerializer,
+                                               OrderDetailSerializer)
+from beauty import signals
+from beauty.utils import ApprovingOrderEmail
 import logging
```
Comment (Serhii-Voloshyn) [inline beauty/api/views.py:28]: 2 blank lines after import section

File: beauty/api/views.py [hunk 10]
```diff
@@ -1,78 +1,101 @@
...
 
         logger.info(f"User {user} was activated")
 
-        return redirect(reverse("api:user-detail", kwargs={"pk": id}))
+        return redirect(reverse("api:user-detail", kwargs={"pk": user_id}))
 
 
 class ResetPasswordView(GenericAPIView):
-    """Generic view for reset password"""
+    """Generic view for reset password."""
```
Comment (Serhii-Voloshyn) [inline beauty/api/views.py:63]: Blank line after class docstring needed

File: beauty/api/views.py [hunk 11]
```diff
@@ -1,78 +1,101 @@
...
+        return redirect(reverse("api:user-detail", kwargs={"pk": user_id}))
 
 
 class CustomUserDetailRUDView(RetrieveUpdateDestroyAPIView):
     """Generic API for users custom GET, PUT and DELETE methods.
-    RUD - Retrieve, Update, Destroy"""
-    # permission_classes = [IsOrReadOnly]
+
+    RUD - Retrieve, Update, Destroy.
+    """
```
Comment (Serhii-Voloshyn) [inline beauty/api/views.py:91]: Blank line after class docstring

File: beauty/beauty/signals.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
+"""Module with all project signals."""
+
+from django.dispatch import Signal, receiver
+from rest_framework.reverse import reverse
+from beauty.utils import StatusOrderEmail
+import logging
+
```
Comment (Serhii-Voloshyn) [inline beauty/beauty/signals.py:7]: 2 blank lines after import section

File: beauty/beauty/tokens.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
+"""Module for all custom project tokens."""
+
+from django.contrib.auth.tokens import PasswordResetTokenGenerator
+import logging
+
```
Comment (Serhii-Voloshyn) [inline beauty/beauty/tokens.py:5]: 2 blank lines after import section

File: beauty/api/models.py [hunk 14]
```diff
@@ -44,7 +43,7 @@ def create_user(self, email: str, first_name: str, password=None,
         user.set_password(password)
         user.save(using=self._db)
 
-        logger.info(f"User {user['first_name']} (id={user.id}) with"
+        logger.info(f"User {user.first_name} (id={user.id}) with"
                     f" {user['email']} was created.")
```
Comment (morento101) [inline beauty/api/models.py:47]: dot notation.....

File: beauty/api/serializers/business_serializers.py [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+        model = Business
+        fields = ('name', 'type', 'owner', 'description')
+
+    def to_representation(self, instance):
```
Comment (pazuzu-ua) [inline beauty/api/serializers/business_serializers.py:10]: add docstring

File: beauty/api/apps.py [hunk 16]
```diff
@@ -4,3 +4,9 @@
 class ApiConfig(AppConfig):
     default_auto_field = 'django.db.models.BigAutoField'
     name = 'api'
+
+    def ready(self):
```
Comment (pazuzu-ua) [inline beauty/api/apps.py:12]: when this method is used?

File: beauty/api/models.py [hunk 17]
```diff
@@ -457,19 +454,24 @@ class Meta:
         verbose_name=_('End time')
     )
     created_at = models.DateTimeField(
-        auto_now=True,
+        auto_now_add=True,
         editable=False,
         verbose_name=_('Created at')
```
Comment (morento101) [inline beauty/api/models.py:468]: As currently implemented, setting auto_now or auto_now_add to True will cause the field to have editable=False and blank=True set.

File: beauty/api/serializers/order_serializers.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+        if specialist == customer:
+            logger.info("Customer and specialist are the same person!")
+
+            raise serializers.ValidationError(
+                {"users": "Customer and specialist are the same person!"})
+        return super().create(validated_data)
+
+
+class OrderDetailSerializer(serializers.ModelSerializer):
```
Comment (pazuzu-ua) [inline beauty/api/serializers/order_serializers.py]: please, rename it to OrderDeleteSerializer for consistency

File: beauty/templates/email/customer_order_status.html [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+{% if order.get_status_display|lower == 'approved' %}
+<p>
+    <i><b>
+        {% trans "Please go to the following page to see order details:" %}
+    </b></i>
+</p>
+<p>
+    <a href="{{ protocol }}://{{ domain }}{{ redirect_url|safe }}">
+        {{ protocol }}://{{ domain }}{{ redirect_url|safe }}
```
Comment (pazuzu-ua) [inline beauty/templates/email/customer_order_status.html:41]: Maybe change to "Click here to confirm" instead of writing a full link?

**Review by pazuzu-ua — APPROVED** [review]

File: beauty/api/tests/factories.py [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+
+    email = factory.LazyAttribute(lambda o: f"{o.first_name.lower()}"
+                                            f"{o.phone_number[-2:]}@example.com")
+    first_name = factory.Faker("first_name")
+    phone_number = factory.Faker("phone_number")
+    password = "1234567890"
+
+    @factory.post_generation
+    def groups(self, create, extracted, **kwargs):
+        """Get the field of ManyToMany for groups."""
```
Comment (morento101) [inline beauty/api/tests/factories.py:47]: Do we still need this method?

File: beauty/api/tests/factories.py [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+        model = Position
+
+    name = factory.Sequence(lambda n: f"Position_{n}")
+    business = factory.SubFactory(BusinessFactory)
+    start_time = factory.LazyFunction(timezone.now)
+    end_time = factory.LazyAttribute(lambda o: o.start_time - timezone.timedelta(hours=8))
+
+    @factory.post_generation
+    def specialist(self, create, extracted, **kwargs):
+        """Get the field of ManyToMany for specialist."""
```
Comment (morento101) [inline beauty/api/tests/factories.py:87]: Do we still need this method?

**Review by morento101 — APPROVED** [review]
Need some minor changes, but ready for merge

File: beauty/api/apps.py (hunk 1 above)
Comment (Misha86) [reply beauty/api/apps.py]: done

File: beauty/api/apps.py (hunk 16 above)
Comment (Misha86) [reply beauty/api/apps.py:12]: Yes. It is from documentation

File: beauty/api/models.py (hunk 17 above)
Comment (Misha86) [reply beauty/api/models.py:468]: done

File: beauty/api/serializers/order_serializers.py (hunk 18 above)
Comment (Misha86) [reply beauty/api/serializers/order_serializers.py]: done

File: beauty/api/tests/test_order_serializers.py (hunk 7 above)
Comment (Misha86) [reply beauty/api/tests/test_order_serializers.py:26]: done

**Review by therobotisnotatoy — APPROVED** [review]

File: beauty/api/tests/test_order_views.py (hunk 8 above)
Comment (Misha86) [reply beauty/api/tests/test_order_views.py:35]: done

File: beauty/api/views.py (hunk 9 above)
Comment (Misha86) [reply beauty/api/views.py:28]: done

File: beauty/api/views.py (hunk 10 above)
Comment (Misha86) [reply beauty/api/views.py:63]: done

File: beauty/api/views.py (hunk 11 above)
Comment (Misha86) [reply beauty/api/views.py:91]: done

File: beauty/beauty/signals.py (hunk 12 above)
Comment (Misha86) [reply beauty/beauty/signals.py:7]: done

File: beauty/beauty/tokens.py (hunk 13 above)
Comment (Misha86) [reply beauty/beauty/tokens.py:5]: done

File: beauty/templates/email/customer_order_status.html (hunk 19 above)
Comment (Misha86) [reply beauty/templates/email/customer_order_status.html:41]: It is link for the customer to redirect to his page

File: beauty/api/apps.py (hunk 2 above)
Comment (Misha86) [reply beauty/api/apps.py]: done

File: beauty/api/models.py (hunk 3 above)
Comment (Misha86) [reply beauty/api/models.py]: done

File: beauty/api/serializers/business_serializers.py (hunk 4 above)
Comment (Misha86) [reply beauty/api/serializers/business_serializers.py:5]: this code from other task, not mine

File: beauty/api/serializers/business_serializers.py (hunk 5 above)
Comment (Misha86) [reply beauty/api/serializers/business_serializers.py:6]: this code from other task, not mine

File: beauty/api/serializers/order_serializers.py (hunk 6 above)
Comment (Misha86) [reply beauty/api/serializers/order_serializers.py:6]: done

File: beauty/api/tests/test_order_serializers.py (hunk 7 above)
Comment (Misha86) [reply beauty/api/tests/test_order_serializers.py:26]: done

File: beauty/api/tests/test_order_views.py (hunk 8 above)
Comment (Misha86) [reply beauty/api/tests/test_order_views.py:35]: done

File: beauty/api/views.py (hunk 9 above)
Comment (Misha86) [reply beauty/api/views.py:28]: done

File: beauty/api/views.py (hunk 10 above)
Comment (Misha86) [reply beauty/api/views.py:63]: done

File: beauty/api/views.py (hunk 11 above)
Comment (Misha86) [reply beauty/api/views.py:91]: done

File: beauty/beauty/signals.py (hunk 12 above)
Comment (Misha86) [reply beauty/beauty/signals.py:7]: done

File: beauty/beauty/tokens.py (hunk 13 above)
Comment (Misha86) [reply beauty/beauty/tokens.py:5]: done

File: beauty/api/models.py (hunk 17 above)
Comment (Misha86) [reply beauty/api/models.py:468]: done

File: beauty/api/serializers/business_serializers.py (hunk 15 above)
Comment (Misha86) [reply beauty/api/serializers/business_serializers.py:10]: this code from other task, not mine(maybe It is done )

File: beauty/api/apps.py (hunk 16 above)
Comment (Misha86) [reply beauty/api/apps.py:12]: It is for connecting our signals. It is from Django documentation

File: beauty/templates/email/customer_order_status.html (hunk 19 above)
Comment (Misha86) [reply beauty/templates/email/customer_order_status.html:41]: It is link for redirect after confirmation

File: beauty/api/tests/factories.py (hunk 20 above)
Comment (Misha86) [reply beauty/api/tests/factories.py:47]: It is for fields ManyToMany from factory boy documentation.

File: beauty/api/tests/factories.py (hunk 21 above)
Comment (Misha86) [reply beauty/api/tests/factories.py:87]: It is for fields ManyToMany from factory boy documentation.

**Pull request merged**